### ISSUE TITLE
Reduction improvements

### ISSUE THREAD

**Issue by lidavidm** [lidavidm/reduct-redux#89]
- [x] Big-step through any previously defined functions
- [ ] Make repeat tail-recursive
- [x] Implement way to expand a name into its definition
- [x] Implement toolbar or overlay to pause/resume execution
- [x] Implement various stepping modes for debugging
- [x] Provide way to cancel reduction in flight
- [ ] Small thing: we should cancel reduction (and stop further evaluation) if `stage.alreadyWon` is true

**Comment by lidavidm** [comment]
Maybe split this into two issues? I kinda made this both reduction UI and reduction semantics by accident

**Comment by EatingW** [comment]
Some of this issue is addressed with code in branch debug-button. There are still bugs in the branch that I'm working on (detailed below). 

Things already implemented:
1. Big-step through any previously defined functions
2. When user clicks on an expression, it will hybrid step (Mode 2), and pause button shows up. 
3. Clicking on pause button takes us to Mode 3, and "play" button and "ffwd" button shows up. 
4. In Mode 3, clicking on the expression does a single step.
5. In Mode 3, clicking on the "play" button begins hybrid step and transitions to Mode 2.
6. In Mode 3, clicking on the "ffwd" button does big step to the final value (sort of, see bugs below).
 
Currently known bugs:
1. After pausing, clicking ">" sometimes get stuck because it thinks the current expression cannot be evaluated for some reason.
2. Clicking ">>" does not take us to the final result but to a function call in the middle (eg. addOne(1))
3. Buttons do not disappear after the expression is evaluated to the end.
4. Sometimes the error sound plays when the expression is evaluated to the end.
5. Clicking an expression with holes, even though cannot be evaluated, will generate buttons. 
6. After an expression steps, the button position does not update with the new expression.
7. Single step still steps into previously defined functions.

**Comment by lidavidm** [comment]
Some style/organization comments:

- Could probably consolidate `debuggerButton` and `button`
- Might be nice to consolidate the `currentlyX` state into a single dictionary if possible; also consider defining methods on stage (`isPaused` etc) like how we have `isHovered` et al

Some comments on bugs:
- Single step not working is because `topExpr` is wrong. I think there is a misunderstanding: `newNodeIds` represents new expressions created, but these aren't necessarily top-level. Thus, you can't always replace the previous topExpr with `newNodeIds[0]`. You are passed `topNodeId`, which is the ID of the top-most node that was actually changed, so you need to check whether `topNodeId` matches current `topNode` before replacing it. (Pushed a fix for this)
- You probably want to generate buttons lazily (i.e. not until update callback is called) or delete buttons if an error is hit (or in the `finally` callback), to prevent extraneous buttons.
- Looks like the `buttonLayout` is coded to track a particular top-level expression? Views aren't carried forward to new expressions, so you'll have to do something more sophisticated there.
- Single step knows nothing about stepping into/out of/etc. You probably want a variation on the hybrid reducer, not "true" single step. (Though, I thought we did want that ability?)
- It all breaks if I undo, then try to click again :stuck_out_tongue: 
- Not sure what's going on with the sound.

**Comment by lidavidm** [comment]
Feedback from instructors:
- Single-step by default so you don't have to rush to click the pause button
- Scrubbable timeline to go forward/back?
- Rewind and play in toolbar
- Global debugger controls? (Avoid questions about concurrency)
- Modes of execution - paused time (single step), normal time (multi step), fast-forward time (big step)
    - Does resumption resume execution on some blocks? All blocks?

**Comment by lidavidm** [comment]
Hold-to-evaluate

Hold-to-preview-definition/slide up to confirm expansion

**Comment by EatingW** [comment]
#111 is where the progress of the global debugger can be tracked.
